File: Python/SinglyLinkedList/SinglyList.py
from Data_Structures.SinglyLinkedList.Node import *
# Data Structures in Python

class SinglyList:
    def __init__(self):
        #Reference to the head node
        self.head = None
        #Number of stack elements
        self.size = 0

    def append(self, element):
        """Add :param element to the back of the list"""
        if self.is_empty():
            self.head = Node(element, self.head)
        else:
            current_node = self.head
            while current_node != None:
                current_node = current_node.next
            current_node.next = Node(element, current_node.next)

    def __len__(self):
        """Return the number of elements in the stack"""
        return self.size

    def is_empty(self):
        """Return True if the stack is empty"""
        return self.size == 0

    def add_first(self, element):
        newest = Node(element, self.head)
        #newest.next = self.head
        self.head = newest
        self.size += 1

    def push(self, element):
        """Add :param element to the top of the stack"""
        #Create and link a new Node
        self.head = Node(element, self.head)
        self.size += 1

    def top(self):
        """Return, but do not remove, the element at the top of the stack
        :except Raise Exception if the stack is empty"""
        if self.is_empty():
            raise TypeError("Stack is empty")
        #Top of stack is at head of the list
        return self.head.element

    def pop(self):
        """Remove and return the element from the top of the stack
        :except Raise Exception if the stack is empty """
        if self.is_empty():
            raise TypeError("Stack is empty")
        answer = self.head.element
        self.head = self.head.next
        self.size -= 1
        return answer

    def remove_first(self):
        if self.head == None:
            raise TypeError("The list is empty")
        self.head = self.head.next
        self.size -= 1

    def get_list(self):
        current_node = self.head
        while current_node != None:
            print(current_node.get_element())
            current_node = current_node.next

    def delete(self, element):
        if self.is_empty():
            raise TypeError("The list is empty")
        current_node = self.head
        while current_node.get_element() != element:
            current_node = current_node.next


"""The space usage is O(n), where n is the current number of elements in the stack."""

File: Java/Arrays/linkedLists/SinglyList.java
package com.interview.questions.array.linkedLists;

public class SinglyList {

    private int size;
    private Node head;
    private Node tail = null;
    private Node currentNode;

    public SinglyList() {
        head = null;
        size = 0;
    }

    /**
     * Add and element to the tail
     * @param element int
     */

    public void append(int element){

        if (isEmpty()) {
            head = new Node(element);
            size++;

        } else {
            currentNode = head;

            while (currentNode.next != null) {
                currentNode = currentNode.next;
            }
            currentNode.next = new Node(element);
            size++;
        }

    }

    /**
     * Return true/false if the list is empty
     * @return boolean
     */
    public boolean isEmpty() {
        return size == 0;
    }

    /**
     * Prints the list from head to tail
     */
    public void getList() {
        Node current = head;
        while (current != null) {
            System.out.println(current.getData());
            current = current.next;
        }
    }

    /**
     * Get the number of elelement in the list
     * @return int size
     */
    public int len(){
        return size;
    }

    public Node deletNode(int toDetele) {
        Node current = head;

        if (current.getData() == toDetele) {
            return head.next;
        }

        while (current.next != null) {
            if (current.next.getData() == toDetele) {
                current.next = current.next.next;
                return head;
            }
            current = current.next;
        }
        return head;
    }

    public void removeDuplicate() {
        Node current = head;
        Node duplicate = null;


    }

    public void kthToLast(int start) {
        int count = 0;
        Node current = head;

        while(current != null) {
            if(count < start) {
                current = current.next;
                count++;
            } else if (count >= start) {
                System.out.println(current.getData());
                current = current.next;
                count++;
            }
        }
    }

    public int recursiveKToLast(Node head, int k) {
        if (head == null) {
            return 0;
        }
        int index = recursiveKToLast(head.next, k) + 1;
        if (index == k) {
            System.out.println(head.getData());
        }
        return index;
    }

}
